agent/config/ratelimited_file_watcher.go

```diff
@@ -0,0 +1,85 @@
+package config
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/go-hclog"
+)
+
+type rateLimitedFileWatcher struct {
+	watcher          Watcher
+	eventCh          chan *FileWatcherEvent
+	coalesceInterval time.Duration
+}
+
+func (r rateLimitedFileWatcher) Start(ctx context.Context) {
+	r.watcher.Start(ctx)
+	r.coalesceTimer(r.watcher.EventsCh(), r.coalesceInterval)
+}
+
+func (r rateLimitedFileWatcher) Stop() error {
+	return r.watcher.Stop()
+}
+
+func (r rateLimitedFileWatcher) Add(filename string) error {
+	return r.watcher.Add(filename)
+}
+
+func (r rateLimitedFileWatcher) Remove(filename string) {
+	r.watcher.Remove(filename)
+}
+
+func (r rateLimitedFileWatcher) Replace(oldFile, newFile string) error {
+	return r.watcher.Replace(oldFile, newFile)
+}
+
+func (r rateLimitedFileWatcher) EventsCh() chan *FileWatcherEvent {
+	return r.eventCh
+}
+
+func NewRateLimitedFileWatcher(configFiles []string, logger hclog.Logger, coalesceInterval time.Duration) (Watcher, error) {
+
+	watcher, err := NewFileWatcher(configFiles, logger)
+	if err != nil {
+		return nil, err
+	}
+	return rateLimitedFileWatcher{
+		watcher:          watcher,
+		coalesceInterval: coalesceInterval,
+		eventCh:          make(chan *FileWatcherEvent),
+	}, nil
+}
+
+func (r rateLimitedFileWatcher) coalesceTimer(inputCh chan *FileWatcherEvent, coalesceDuration time.Duration) {
+	var coalesceTimer *time.Timer = nil
+	sendCh := make(chan struct{})
+	FileWatcherEvents := make([]string, 0)
+	go func() {
+		for {
+			select {
+			case event, ok := <-inputCh:
```

Comment:
Do you need an explicit `ctx.Done()` select case here, or should the closing of `inputCh` be enough to shut things down?

Reply:
The channel close should be enough. we for range on the channel and if it's closed we leave the loop.